lib/enum.rb

```diff
@@ -0,0 +1,108 @@
+# frozen_string_literal: true
+
+# !/home/eduardo/.rvm/rubies/ruby-2.6.3/bin/ruby
+
+module Enumerable
+  def my_each
+    return to_enum unless block_given?
+
+    to_a.length.times { |i| yield(to_a[i]) }
+    self
+  end
+
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    to_a.length.times { |i| yield(to_a[i], i) }
+    self
+  end
+
+  def my_select
+    return to_enum unless block_given?
+
+    arr = []
+    my_each { |i| arr.push(i) if yield(i) }
+    arr
+  end
+
+  def my_all?(pattern = nil)
+    if block_given?
+      my_each { |elem| return false unless yield(elem) }
+    elsif pattern
+      my_each { |elem| return false unless verify?(elem, pattern) }
+    else
+      my_each { |elem| return false unless elem }
+    end
+    true
+  end
+
+  def my_any?(pattern = nil)
+    if block_given?
+      my_each { |elem| return true if yield(elem) }
+    elsif pattern
+      my_each { |elem| return true if verify?(elem, pattern) }
+    else
+      my_each { |elem| return true if elem }
+    end
+    false
+  end
+
+  def my_none?(pattern = nil)
+    if block_given?
+      my_each { |elem| return false if yield(elem) }
+    elsif pattern
+      my_each { |elem| return false if verify?(elem, pattern) }
+    else
+      my_each { |elem| return false if elem }
+    end
+    true
+  end
```

Comment:
* [ ] my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern